extmod/foundation-rust/embedded-storage-redundant/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+
+//! Embedded storage redundant reader.
+//!
+//! This crate provides a layer to a [`embedded-storage`] NOR flash device to
+//! perform multiple redundant reads to the storage device while comparing
+//! the contents of each read for consistency.  If the contents don't match
+//! then it is retried a certain amount of times before giving up and
+//! returning the first buffer.
+//!
+//! The [`RedundantRead`] structure accepts parameter for the buffer size,
+//! number of redundant copies and retries.
+//!
+//! The higher the buffer size and more copies the higher the time to
+//! compare the buffers.
+
+#![no_std]
+
+use embedded_storage::nor_flash::{ErrorType, NorFlash, ReadNorFlash};
+use heapless::Vec;
+
+/// Redundant reads layer.
+///
+/// - `N` is the length of each redundant copy.
+/// - `COPIES` the number of redundant copies.
+#[derive(Debug)]
+pub struct RedundantRead<S, const N: usize, const COPIES: usize> {
+    storage: S,
+    buffers: Vec<Vec<u8, N>, COPIES>,
+    retries: usize,
+}
+
+impl<S, const N: usize, const COPIES: usize> RedundantRead<S, N, COPIES> {
+    pub fn new(storage: S, retries: usize) -> Self {
+        let mut buffers = Vec::new();
+        for _ in 0..COPIES {
+            buffers.push(Vec::new()).unwrap();
+        }
+
+        Self {
+            storage,
+            buffers,
+            retries,
+        }
+    }
+
+    /// Return the inner storage as a reference.
+    pub fn as_inner(&self) -> &S {
+        &self.storage
+    }
+
+    /// Return the inner storage as a mutable reference.
+    pub fn as_mut_inner(&mut self) -> &mut S {
+        &mut self.storage
+    }
+}
+
+impl<S: ReadNorFlash, const N: usize, const COPIES: usize>
+    RedundantRead<S, N, COPIES>
+{
+    fn read_copies(&mut self, offset: u32, len: usize) -> Result<(), S::Error> {
+        for i in 0..COPIES {
+            let buffer = &mut self.buffers[i];
+            buffer.clear();
+            buffer
+                .resize(len, 0)
+                .expect("caller pre-condition prevents this");
+
+            self.storage.read(offset, buffer)?;
+        }
+
+        Ok(())
+    }
+
+    fn are_buffers_equal(&self) -> bool {
+        for i in 0..(COPIES - 1) {
+            if self.buffers[i] != self.buffers[i + 1] {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<S: ErrorType, const N: usize, const COPIES: usize> ErrorType
+    for RedundantRead<S, N, COPIES>
+{
+    type Error = S::Error;
+}
+
+impl<S: ReadNorFlash, const N: usize, const COPIES: usize> ReadNorFlash
+    for RedundantRead<S, N, COPIES>
+{
+    const READ_SIZE: usize = S::READ_SIZE;
+
+    fn read(
+        &mut self,
+        offset: u32,
+        bytes: &mut [u8],
+    ) -> Result<(), Self::Error> {
+        // Read is larger than what we support, just do it without redundancy.
+        if bytes.len() >= N {
+            return self.storage.read(offset, bytes);
+        }
+
+        for _ in 0..self.retries {
```

Comment:
This is different that what I implemented for Gen 2.  For Gen 2, I would read all the buffers once, and then if there was a discrepancy, it just iterates through the buffer copies one by one to reread them and then compares after each subsequent read.  The idea was to try to minimize the number of repeated reads if, for example, just one buffer differed.

With the loop as written here, you will always reread every buffer, which will be slower.